types/recompose/index.d.ts

```diff
@@ -268,20 +269,22 @@ declare module 'recompose' {
 
     // Static property helpers: https://github.com/acdlite/recompose/blob/master/docs/API.md#static-property-helpers
 
+    type ComponentModifier = <T extends Component>(component: T) => T;
+
     // setStatic: https://github.com/acdlite/recompose/blob/master/docs/API.md#setStatic
-    export function setStatic<TOutter>(
+    export function setStatic(
```

Comment:
If any caller used to specify a type argument, they will be broken by this.

Reply:
(Consider `setStatic<P = {}>` and `ComponentModifier<P>`.)

Reply:
Hey @amcasey, thanks for taking a look at this.

Yes, it is a breaking change, the typing for `setStatic` should not take a type argument.

`setStatic` creates a function that can be called on *any* component.  It sets a static property on the component and returns it, [that's literally all it does](https://github.com/acdlite/recompose/blob/5b8498a48e8f7ad9dd6b2706be2fc3311d86543c/src/packages/recompose/setStatic.js#L3-L5).

Requiring a type argument means that the resulting function is incorrectly typed as only being able to accept a component with props matching the type that was passed as the type argument when it was created with `setStatic`.

In addition to artificially limiting the resulting function, the type argument also requires the user to know the props of the component they are planning on calling the function with and to accurately pass the type of those props when creating the function using `setStatic`.  This creates unnecessary work and introduces the possibility for errors and confusion if the type argument is not passed correctly.  (I've answered questions on SO about this).

This change corrects the typing for `setStatic` which also simplifies its use.

I bumped the version, is there anything else that needs to be done for a breaking change like this?

Reply:
I trust your judgement (and mostly understand your explanation) that setStatic *shouldn't* be generic.  However, this appears to be quite a widely used package and so some weight has to be given to call patterns that exist now.  Personally, I think it would be very helpful to have input from an additional reviewer who knows more about the domain, but I'm not sure how to find such a person.  In the interim, I think bumping the version is probably a sufficient mitigation.

Reply:
Anyone who was using the older incorrect typing will have code that looks something like this:

`setStatic<IProps>('prop', 'value')`

When they update to the new typing they will see an error for that line like this:

`[ts] Expected 0 type arguments, but got 1.`

They will simply need to remove the type argument so the line looks like this:

`setStatic('prop', 'value')`

The same process and result applies to any previous uses of `setDisplayName`.

This is an easy code migration and the benefit is correct typing.  An additional benefit is that anyone using `recompose` for the first time won't need to worry about the unnecessary type arguments for `setStatic` and `setDisplayName` and will get correct typing.